consensus/src/marshal/actor.rs

```diff
@@ -0,0 +1,764 @@
+
+        // Process all finalized blocks in order (fetching any that are missing)
+        let (mut notifier_tx, notifier_rx) = mpsc::channel::<()>(1);
+        let (orchestrator_sender, mut orchestrator_receiver) = mpsc::channel(2); // buffer to send processed while moving forward
+        let orchestrator = Orchestrator::new(orchestrator_sender);
+        let finalizer = Finalizer::new(
+            self.context.with_label("finalizer"),
+            self.partition_prefix.clone(),
+            application,
+            orchestrator,
+            notifier_rx,
+        )
+        .await;
+        self.context
+            .with_label("finalizer")
+            .spawn(|_| finalizer.run());
+
+        // Handle messages
+        loop {
+            // Select messages
+            select! {
+                // Handle consensus before finalizer or backfiller
+                mailbox_message = self.mailbox.next() => {
+                    let Some(message) = mailbox_message else {
+                        info!("Mailbox closed, shutting down");
+                        return;
+                    };
+                    match message {
+                        Message::Broadcast { payload } => {
+                            let ack = buffer.broadcast(Recipients::All, payload).await;
+                            drop(ack);
+                        }
+                        Message::Verified { view, payload } => {
+                            self.put_verified(view, payload.commitment(), payload).await;
+                        }
+                        Message::Notarization { notarization } => {
+                            let view = notarization.proposal.view;
+                            let commitment = notarization.proposal.payload;
+
+                            // If found, store notarization
+                            if let Some(block) = self.search_for_block(&mut buffer, commitment, SearchDepth::Verified).await {
+                                self.put_notarization(view, commitment, notarization, block).await;
+                                continue;
+                            }
+
+                            // Fetch from network
+                            //
+                            // We don't worry about retaining the proof because any peer must provide
+                            // it to us when serving the notarization.
+                            debug!(view, "notarized block missing");
+                            resolver_by_view.fetch(view.into()).await;
+                        }
+                        Message::Finalization { finalization } => {
+                            let view = finalization.proposal.view;
+                            let commitment = finalization.proposal.payload;
+
+                            // If found, store finalization
+                            if let Some(block) = self.search_for_block(&mut buffer, commitment, SearchDepth::Notarized).await {
```

Comment:
I'm tempted to suggest we store finalization both by view and height (pruning the finalization by view when we prune the other stores?).

As discussed, it seems suboptimal to drop finalizations here (that's my b lol)...